### PR TITLE
Addresses several shortcomings in current worflows (mainly mainnet related)

### PR DESCRIPTION
**Pull request by d10r**
See https://github.com/superfluid-finance/protocol-monorepo/pull/905#issuecomment-1162820427

### CODE REVIEW

**Comment by d10r** [comment]
piggy-backed some stuff here: fix to pure super token deploy script, viz of a CFA operation

The deploy script is now single-network, can also be used for testnets, can specify an RPC override (added to truffle-config).
Idea for handling verification incl. security improvements:
* let the deploy script upload addresses of newly deployed contracts somewhere (similar to coverage uploads?)
* let a script running on another machine download those address lists and verify against the same contract version

What this achieves:
* if the machine deploying contracts is compromised, verification will fail
* the contracts can still be verified before asking gov to sign, allowing gov to verify
* can easily add more verification methods (e.g. sourcify) and triggers for when it's done, e.g. notify gov to sign, with a list of new contracts and links to the verified code

We may need to change the verification script to directly verify logic contracts instead of indirectly through the proxy. Since proxy verification is still partly broken anyway, we may look for a way to revert the plugin behavior to not be proxy-aware, e.g. by forking it.

**Comment by d10r** [comment]
_Even just seeing which RPC was used would already be helpful_ is still unsolved.  
I tried to add a console.log of the RPC URL to the deploy script, but web3.js seems to have no interface for getting it. In a local attempt, `web3.currentProvider.engine._providers[3].rpcUrl` would do the job, but that looks like some internal API which may break anytime.

In general, I'm not happy with the status quo.
Workflow runs are ok for automated background tasks, but too slow for manual invocation like listings. Shouldn't take minutes just to set up. Although more caching etc. may help, I'm becoming increasingly skeptical that GH actions is the right tool, will further contemplate possible alternatives.

The fragility of the upgrade process is also becoming a serious issue with expensive networks. Having local state from which consecutive runs can resume could help.

As for verification, I think we could fully automate it by having the deploy script publish a list of deployed contract addresses (e.g. as created by `info-print-contract-addresses.js`) which is picked up and fed to the verification script.  
If this is running in an distinct context, it would also be a security enhancement, offering protection against the deploy process being compromised. Verification would then fail, highlighting a potential issue before conducting the actual upgrade of the proxies. The multisig signers manually checking the code via Explorers isn't realistic anyway (and wouldn't be effective either if it were).

**Comment by hellwolf** [comment]
@d10r  self-hosted running could be another option for protected environment for deployment.

image


But let's clarify the problem you are trying to address:

1. "but too slow for manual invocation like listings"

Is that the only issue or is there anything else?


For that particular problem (1), there are cache solutions too:

https://github.community/t/how-to-setup-yarn-install-cache-on-github-action/18138/2
https://github.com/actions/cache

Overall I think it's a general immutable build output problem. With custom solution, we would either not solving it systematically, or re-solve the same problem; while risking forgetting the benefits from an existing workflow system.

File: .github/workflows/handler.placeholder-for-testing.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Raison d'être: new handlers can't be tested before being merged to dev.
```
Comment (hellwolf) [inline .github/workflows/handler.placeholder-for-testing.yml:1]: good sneaky way.

Also, github allows certain secrets only be available to certain branches through "environment" secrets container, which is a good security measure for our sized projects when we have many collaborators and various computer security setup


image

File: packages/ethereum-contracts/package.json [hunk 2]
```diff
@@ -75,6 +75,6 @@
         "solidity-coverage": "0.7.21",
         "solidity-docgen": "^0.6.0-beta.17",
         "truffle-flattener": "^1.6.0",
-        "truffle-plugin-verify": "^0.5.26"
+        "truffle-plugin-verify": "d10r/truffle-plugin-verify#verifyProxyFixed"
```
Comment (hellwolf) [inline packages/ethereum-contracts/package.json:78]: :(

**Comment by hellwolf** [comment]
```
$ fgrep 'npx truffle' .github/ packages/ethereum-contracts/{scripts,tasks} -R | grep -v 'Usage' | grep -v -e 'truffle (exec|run)'
.github/workflows/handler.list-super-token.yml:          npx truffle \
.github/workflows/handler.deploy-to-network.yml:          npx truffle exec --network ${{ github.event.inputs.network }} scripts/deploy-framework.js
.github/workflows/handler.deploy-to-testnets.yml:          npx truffle exec --network ${{ matrix.network }} scripts/deploy-test-environment.js
.github/workflows/handler.deploy-to-testnets.yml:          npx truffle exec --network ${{ matrix.network }} scripts/info-print-contract-addresses.js : addresses.vars
.github/workflows/handler.run-ethereum-contracts-script.yml:          npx truffle \
.github/workflows/handler.verify-contracts.yml:          npx truffle exec --network ${{ matrix.network }} scripts/info-print-contract-addresses.js : addresses.vars
.github/workflows/ci.canary.yml:          npx truffle exec --network ${{ matrix.network }} scripts/deploy-test-environment.js
packages/ethereum-contracts/scripts/gov-create-new-app-registration-key.js: * npx truffle exec scripts/gov-create-new-app-registration-key.js : {DEPLOYER} {REGISTRATION_KEY} [EXPIRATION_TS]
packages/ethereum-contracts/tasks/etherscan-verify-framework.sh:    npx truffle run --network $TRUFFLE_NETWORK verify "$@"
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${SUPERFLUID_HOST_PROXY}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${SUPERFLUID_SUPER_TOKEN_FACTORY_PROXY}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${CFA_PROXY}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${IDA_PROXY}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:    npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${SUPER_TOKEN_FDAIX}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:    npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${SUPER_TOKEN_FUSDCX}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:    npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy UUPSProxy@${SUPER_TOKEN_FTUSDX}
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:    npx truffle --network $TRUFFLE_NETWORK run verify --verify-proxy SETHProxy@${SUPER_TOKEN_NATIVE_COIN}
```

**Comment by hellwolf** [comment]
Fixed:

```
$ fgrep 'npx truffle' .github/ packages/ethereum-contracts/{scripts,tasks} -R | grep -v 'Usage' | grep -v -E 'truffle (exec|run)'
packages/ethereum-contracts/tasks/etherscan-verify-proxies.sh:echo TRUFFLE_RUN_VERIFY="npx truffle --network $TRUFFLE_NETWORK run verify"
```

**Review by hellwolf — APPROVED** [review]

**Pull request merged**